### DISCUSSION TITLE
Understanding the power curve

### DISCUSSION THREAD

**Discussion by moptis** [NatLabRockies/floris#467] [Q&A] [answered]
I'm a newbie to Floris, so this answer must be very obvious. But how does a 15 MW IEA turbine, for example, have a peak power value of 0.45? Is this some scaled density value? I can't find any documentation that address this.

Thanks!

**Comment by moptis** [comment, marked as answer]
OK, just figured out that it's a power coefficient curve. Next question though: how is total power production at the turbine determined? There doesn't appear to be any clear indication of a 5 MW capacity in the NREL 5MW data, for example.

**Reply by rafmudaf** [reply]
Hi Mike,

Each turbine's power is calculated in the [Turbine.power](https://github.com/NREL/floris/blob/main/floris/simulation/turbine.py#L80) function for each wind condition. The gist of that function is that the power comes almost directly from the wind speed at the turbine.

As for evaluating the power of the 5MW, try this simple test and be sure to put your script in the `examples/` directory:

```python
from floris.tools import FlorisInterface

fi = FlorisInterface("inputs/gch.yaml")

wind_speeds = [float(i) for i in range(4,25)]

fi.reinitialize( layout=( [0.0], [0.0] ) )
fi.reinitialize(wind_directions=[270.], wind_speeds=wind_speeds)
fi.calculate_wake()
turbine_powers = fi.get_turbine_powers()/1000.

print(f"| Wind Speed | Power (kW) |")
for i in range(len(wind_speeds)):
    print(f"| {wind_speeds[i]:10.1f} | {turbine_powers[0,i,0]:10.1f} |")
```

You can switch the turbine type in the input file and compare their calculated powers, as well.

Its good to hear from you!

Raf

**Reply by rafmudaf** [reply]
In case its not clear, the FlorisInterface is meant to be a sort of user API that provides a few helper routines to do common things like calculate AEP and get the turbine powers. Everything in `floris.tools` is meant to be more user-facing. The power calculation lives in `floris.simulation` and it shouldn't really be accessed directly from your scripts. It's better to use the functions in `floris.tools` if they're available.

I know the documentation is lacking, and this is something we are working to address. Already your question has shed some light on what we need to add to the documentation, but if you have additional feedback please comment in the [discussion on documentation](https://github.com/NREL/floris/discussions/468).

**Reply by moptis** [reply]
Thanks Raphael, good to hear from you too! This is all very helpful.

So for the 5 MW example, where is the turbine capacity actually specified, i.e., the 5 MW? I'd like to create my own turbine type and can create the power and thrust coefficient curves easy enough. But it's not clear how to assign total turbine capacity.

For what it's worth, from an industry perspective, I'm interested in using FLORIS to find optimal offshore layouts. I'd be happy to keep a thread going on my experiences and where I think more documentation might be useful.

Mike

**Reply by rafmudaf** [reply]
The turbine capacity is implicit in the power curve. These are generally gotten from a tool like OpenFAST. That's why the power curve flatlines at 5MW, for example. There are no turbine controls in FLORIS (like blade pitch control), but there is a controller in OpenFAST that holds the power at rated in region 3. So you won't find a place in floris to enter "5MW", but that information should be included in the Cp/Ct vs wind speed table.

**Reply by rafmudaf** [reply]
> For what it's worth, from an industry perspective, I'm interested in using FLORIS to find optimal offshore layouts. I'd be happy to keep a thread going on my experiences and where I think more documentation might be useful.

Please do stay in touch on that. We have similar efforts at NREL, so if want to discuss in more detail just get in touch. My email address is on my GitHub profile (github.com/rafmudaf).

**Reply by moptis** [reply]
Ah ok, makes sense, thanks Raphael. 

My confusion came from the use of Cp data in the turbine library rather than actual power output in kilowatts. I'm just not used to seeing Cp. Turbine manufacturers always provide us power output and Ct data and we're generally left to calculate Cp ourselves. 

I'm good to go from here. Thanks again.

**Reply by paulf81** [reply]
Thanks for this discussion, just to add a few quick comments of my own here:
- Make sure Cp is "electrical" Cp (in the FAST terminology) meaning it has shaft and generator efficiencies "baked in" and the final output is for example, the exact rated power you expect
- I've found the example 15_check_turbine.py really useful in doing a last sanity check on your input file
- I've gone through this process a few times (mocking up a FLORIS input of a new turbine given just a few descriptors, adapting from the 5MW or 15MW for example), maybe some included function that automates this would help going forward?
